WordPress/gutenberg
# Tree-shaking block styles on the frontend

## What?
In https://github.com/WordPress/gutenberg/pull/25220 we split the CSS files for blocks, so block-themes can load separate files per-block, depending on what blocks are present on a page.
This PR takes the concept further, and allows loading CSS files depending on the actual block contents/classes etc.

## Why?
Block stylesheets contain all styles for all scenarios a block may encounter. Some of those scenarios are not that frequent, yet their styles must always be loaded because we currently have no way to differentiate what's used and what's not.
This results is loading way more CSS than what is actually needed in a page, wasting resources.
The implementation in this PR allows us to do some tree-shaking and only load the styles that are actually used on a page.
The benefits will be improved performance for site visitors - especially for block themes, and improved sustainability for the WP ecosystem.

## How?
The PR adds a `wp_maybe_inline_block_style_parts` function. That function checks for a `style.[].parts` (array) entry in `block.json` files (see example in the included paragraph block tweaks)
It then adds a filter on `render_block`, and if the block contains the parts from the above array, it includes the corresponding CSS file (named `styles/{$part}.css`).

This PR also includes changes to the `core/paragraph` block, splitting it to separate files, to make the implementation clearer and easier to test.

## Testing Instructions

Add a paragraph block on a page, load the post on the frontend and ensure that if you don't use a dropcap, the CSS for it doesn't get loaded. If you edit the post and add a dropcap, the CSS for it should then become available on the frontend.
This change should not have any impact on the editor, this is a frontend-only improvement.
